hardware_interface/doc/writing_new_hardware_interface.rst

```diff
@@ -0,0 +1,139 @@
+   Optionally add ``visibility_control.h`` with the definition of export rules for Windows.
+   You can copy this file from an existing controller package and change the name prefix to the ``<PACKAGE_NAME>``.
+
+3. **Adding declarations into header file (.hpp)**
+
+   1. Take care that you use header guards. ROS2-style is using ``#ifndef`` and ``#define`` preprocessor directives. (For more information on this, a search engine is your friend :) ).
+
+   2. Include ``"hardware_interface/$interface_type$_interface.hpp"`` and ``visibility_control.h`` if you are using one.
+      ``$interface_type$`` can be ``Actutor``, ``Sensor`` or ``System`` depending on the type of hardware you are using. for more details about each type check `Hardware Components description <https://ros-controls.github.io/control.ros.org/getting_started.html#hardware-components>`_.
+
+   3. Define a unique namespace for your hardware_interface. This is usually a package name written in ``snake_case``.
+
+   4. Define the class of the hardware_interface, extending ``$InterfaceType$Interface``, e.g.,
+      .. code:: c++
+      class HardwareInterfaceName : public hardware_interface::$InterfaceType$Interface
+
+   5. Add a constructor without parameters and the following public methods implementing ``LifecycleNodeInterface``: ``on_configure``, ``on_cleanup``, ``on_shutdown``, ``on_activate``, ``on_deactivate``, ``on_error``; and overriding ``$InterfaceType$Interface`` definition: ``on_init``, ``export_state_interfaces``, ``export_command_interfaces``, ``prepare_command_mode_switch`` (optional), ``perform_command_mode_switch`` (optional), ``read``, ``write``.
+   For further explanation of hardware-lifecycle check the `pending pull request <https://github.com/ros-controls/ros2_control/pull/559/files#diff-2bd171d85b028c1b15b03b27d4e6dcbb87e52f705042bf111840e7a28ab268fc>`_ and for exact definitions of methods check the ``"hardware_interface/$interface_type$_interface.hpp"`` header or `doxygen documentation <http://control.ros.org/api/namespacehardware__interface.html>`_ for *Actuator*, *Sensor* or *System*.
+
+4. **Adding definitions into source file (.cpp)**
+
+   1. Include the header file of your hardware interface and add a namespace definition to simplify further development.
+
+   2. Implement ``on_init`` method. There, you should all initialize member variables and process hardware's parameters from ``info`` argument.
+      In the first line usually the parents ``on_init`` is called to process standard values, like name. This is done using: ``hardware_interface::(Actuator|Sensor|System)Interface::on_init(info)``.
+      If all required parameters are set and valid and everything works fine return ``CallbackReturn::SUCCESS`` or ``return CallbackReturn::ERROR`` otherwise.
+
+   4. Write the ``on_configure`` method where you usually setup the communication to the hardware and set everything up so that the hardware can be activated.
```

Comment:
What do you think about this?

```suggestion
   4. Write the ``on_configure`` method where you usually setup the communication to the hardware, allocate the memory and set everything up so that the hardware can be activated.
```